core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -356,6 +356,90 @@ public void write(byte[] b, int off, int len) throws IOException {
         writeBytes(b, off, len);
     }
 
+    /**
+     * Uses {@link StreamOutput#writeVInt(int)} to first provide the size, then it writes the array using each instances
+     * {@link Writeable#writeTo(StreamOutput)} method.
+     * <p>
+     * This is meant to replace boilerplate surrounding code of the form:
+     * <pre>
+     * YourObject[] objects = ...;
+     * out.writeVInt(objects.length);
+     * for (YourObject object : objects) {
+     *     object.writeTo(out);
+     * }
+     * </pre>
+     *
+     * @param array The array to write
+     */
+    public <T extends Writeable> void writeArray(T[] array) throws IOException {
+        writeVInt(array.length);
+        for (T t : array) {
+            t.writeTo(this);
+        }
+    }
+
+    /**
+     * Writes a {@link #writeBoolean(boolean) boolean} {@code true} if the array is non-{@code null}. Otherwise {@code false}.
+     * <p>
+     * This is meant to replace boilerplate surrounding code of the form:
+     * <pre>
+     * YourObject[] objects = ...;
+     * if (objects != null) {
+     *     out.writeBoolean(true);
+     *     out.writeVInt(objects.length);
+     *     for (YourObject object : objects) {
+     *         object.writeTo(out);
+     *     }
+     * } else {
+     *     out.writeBoolean(false);
+     * }
+     * </pre>
+     *
+     * @param array The array to write
+     */
+    public <T extends Writeable> void writeOptionalArray(T[] array) throws IOException {
+        if (array == null) {
+            writeBoolean(false);
+        } else {
+            writeBoolean(true);
+            writeArray(array);
+        }
+    }
+
+    /**
+     * This is an identical copy of {@link #writeArray(Writeable[])}, except it accepts {@link Streamable[]}, which has an identical
+     * function, but a different meaning from {@link Writeable} ({@code writeTo(StreamOutput)}).
+     * <p>
+     * New code should use {@link Writeable} and its contract to create immutable objects rather than {@link Streamable}.
+     *
+     * @param array The array to write
+     * @see #writeArray(Writeable[])
+     */
+    public <T extends Streamable> void writeArray(T[] array) throws IOException {
+        writeVInt(array.length);
```

Comment:
I didn't want to have Streamable extend Writeable because I think it clouds the meaning of Writeable in those cases (its intent of immutability is suddenly lost when used with Streamable). It sounds like I'm going to shift to `List` from arrays anyway, and thereby drop these methods altogether.